pkg/querier/frontend/frontend.go

```diff
@@ -0,0 +1,243 @@
+package frontend
+
+import (
+	"context"
+	"flag"
+	"math/rand"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/weaveworks/common/httpgrpc"
+	"github.com/weaveworks/common/httpgrpc/server"
+	"github.com/weaveworks/common/user"
+)
+
+var (
+	queueDutation = promauto.NewHistogram(prometheus.HistogramOpts{
+		Namespace: "cortex",
+		Name:      "query_frontend_queue_duration_seconds",
+		Help:      "Time spend by requests queued.",
+		Buckets:   prometheus.DefBuckets,
+	})
+	retries = promauto.NewHistogram(prometheus.HistogramOpts{
+		Namespace: "cortex",
+		Name:      "query_frontend_retries",
+		Help:      "Number of times a request is retried.",
+		Buckets:   []float64{0, 1, 2, 3, 4, 5},
+	})
+	queueLength = promauto.NewGauge(prometheus.GaugeOpts{
+		Namespace: "cortex",
+		Name:      "query_frontend_queue_length",
+		Help:      "Number of queries in the queue.",
+	})
+
+	errServerClosing  = httpgrpc.Errorf(http.StatusTeapot, "server closing down")
```

Comment:
Nice status code

Reply:
I was hoping for a official "closing down" status code, but I couldn't find anything...